File: go/algorithms/tree.go
package algorithms

// TreeSort returns a new sorted slice using a binary search tree.
func TreeSort(arr []int) []int {
    var root *treeNode
    for _, v := range arr {
        root = insertNode(root, v)
    }
    out := make([]int, 0, len(arr))
    inorder(root, &out)
    return out
}

type treeNode struct {
    val   int
    left  *treeNode
    right *treeNode
}

func insertNode(n *treeNode, v int) *treeNode {
    if n == nil {
        return &treeNode{val: v}
    }
    if v < n.val {
        n.left = insertNode(n.left, v)
    } else {
        n.right = insertNode(n.right, v)
    }
    return n
}

func inorder(n *treeNode, out *[]int) {
    if n == nil {
        return
    }
    inorder(n.left, out)
    *out = append(*out, n.val)
    inorder(n.right, out)
}

File: rust/src/tree.rs
pub fn tree_sort(arr: &[i32]) -> Vec<i32> {
    let mut root: Option<Box<TreeNode>> = None;
    for &v in arr.iter() {
        root = insert(root, v);
    }
    let mut out = Vec::with_capacity(arr.len());
    inorder(&root, &mut out);
    out
}

struct TreeNode {
    val: i32,
    left: Option<Box<TreeNode>>,
    right: Option<Box<TreeNode>>,
}

fn insert(node: Option<Box<TreeNode>>, v: i32) -> Option<Box<TreeNode>> {
    match node {
        None => Some(Box::new(TreeNode { val: v, left: None, right: None })),
        Some(mut n) => {
            if v < n.val {
                n.left = insert(n.left.take(), v);
            } else {
                n.right = insert(n.right.take(), v);
            }
            Some(n)
        }
    }
}

fn inorder(node: &Option<Box<TreeNode>>, out: &mut Vec<i32>) {
    if let Some(n) = node {
        inorder(&n.left, out);
        out.push(n.val);
        inorder(&n.right, out);
    }
}
